### PR TITLE
Debug: add build branch to title

### PR DESCRIPTION
**Pull request by gerhardol**
## Proposed changes

Add the branch the PR is built from to the window title.
The commit sha is visible in About, but especially if the branch is rebased there are several extra steps to find the branch (and the intended functionality included.

An alternative is to add the branch name in About for 'debug builds.

### Before
image

### After

image

## Test methodology 

Manual

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

### CODE REVIEW

File: GitCommands/AppTitleGenerator.cs [hunk 1]
```diff
@@ -46,10 +50,17 @@ public string Generate(string workingDir = null, bool isValidWorkingDir = false,
             var description = _description.Get(workingDir);
 
 #if DEBUG
-            return $"{description} ({branchName}) - Git Extensions [DEBUG]";
+            return $"{description} ({branchName}) - Git Extensions [{_buildBranch}]";
```
Comment (mstv) [inline GitCommands/AppTitleGenerator.cs]:
```suggestion
            _buildBranch = _buildBranch ?? "DEBUG";
            return $"{description} ({branchName}) - Git Extensions [{_buildBranch}]";
```
would be bulletproof.

File: GitCommands/AppTitleGenerator.cs [hunk 2]
```diff
@@ -46,10 +50,17 @@ public string Generate(string workingDir = null, bool isValidWorkingDir = false,
...
 #else
             return $"{description} ({branchName}) - Git Extensions";
+#endif
+        }
+
+        public static void Initialise(string buildBranch)
+        {
+#if DEBUG
+            _buildBranch = buildBranch;
```
Comment (mstv) [inline GitCommands/AppTitleGenerator.cs]: Turning this into a property would improve readability.

Comment (gerhardol) [reply GitCommands/AppTitleGenerator.cs]: done. I followed the pattern in UserEnvironmentInformation. (no static field either)

File: GitCommands/AppTitleGenerator.cs [hunk 3]
```diff
@@ -46,10 +50,17 @@ public string Generate(string workingDir = null, bool isValidWorkingDir = false,
             var description = _description.Get(workingDir);
 
 #if DEBUG
-            return $"{description} ({branchName}) - Git Extensions [DEBUG]";
+            return $"{description} ({branchName}) - Git Extensions [{BuildBranch ?? "DEBUG"}]";
```
Comment (mstv) [inline GitCommands/AppTitleGenerator.cs]: I am afraid this will not compile.

File: GitCommands/AppTitleGenerator.cs [hunk 4]
```diff
@@ -46,10 +50,17 @@ public string Generate(string workingDir = null, bool isValidWorkingDir = false,
...
 #else
             return $"{description} ({branchName}) - Git Extensions";
+#endif
+        }
+
+        public static void Initialise(string buildBranch)
```
Comment (mstv) [inline GitCommands/AppTitleGenerator.cs]: I proposed the property in order to replace this method.

File: GitCommands/AppTitleGenerator.cs (hunk 3 above)
Comment (gerhardol) [reply GitCommands/AppTitleGenerator.cs]: This works - if you request changes is another thing

File: GitCommands/AppTitleGenerator.cs (hunk 4 above)
Comment (gerhardol) [reply GitCommands/AppTitleGenerator.cs]: If this is changed, the #if DEBUG need to be handled in Program.cs too - is that requested?

File: GitCommands/AppTitleGenerator.cs (hunk 3 above)
Comment (mstv) [reply GitCommands/AppTitleGenerator.cs]: Perfect if it works. I had problems with similar constructs before.

File: GitCommands/AppTitleGenerator.cs (hunk 4 above)
Comment (mstv) [reply GitCommands/AppTitleGenerator.cs]: I'd make the property independent from `DEBUG`. Then it will just not be used. @RussKie?

Comment (RussKie) [reply GitCommands/AppTitleGenerator.cs]: I'd like to ponder on this a little. I suspect we can do it slightly differently, but I need some time to investigate.

Comment (gerhardol) [reply GitCommands/AppTitleGenerator.cs]: Fine, I just seen some needs to track my private builds

**Comment by vbjay** [comment]
Instead of title could be in about dialog.

**Comment by gerhardol** [comment]
> Instead of title could be in about dialog.

This alternative is in the Description.
I just want something, if About can be agreed on instead, I will change.

**Comment by mstv** [comment]
It affects only the debug builds. It should be easy for developers to distinguish the running instances.
IMO it's a very good idea to replace the constant string "DEBUG" with something more informative.

**Comment by pmiossec** [comment]
I also agree. The added value is to see it immediately without efforts. So, title is a good place.

**Comment by gerhardol** [comment]
Any reviews, to merge or dismiss this PR?
We have too many PRs open right now...

File: GitCommands/AppTitleGenerator.cs (hunk 4 above)
Comment (mstv) [reply GitCommands/AppTitleGenerator.cs]: > Any reviews, to merge or dismiss this PR?

The private property `BuildBranch` is pointless IMO. (I don't know which is a better English word if this was too harsh.)
Either a non-`DEBUG` public propery `BuildBranch` or `Initialise`. I prefer the first for readability.

**Comment by RussKie** [comment]
I'll look at it tonight

**Comment by RussKie** [comment]
I'm proposing to show the following:
image

If there is no local branch but there's a remote, it will show as:
image

If there is no branches at all, it will show some git magic:
image


A release build remains unaffected.

File: GitCommands/AppTitleGenerator.cs (hunk 4 above)
Comment (mstv) [reply GitCommands/AppTitleGenerator.cs]: @RussKie, what do you think about non-`DEBUG` public propery `BuildBranch` vs. `Initialise`?

Comment (RussKie) [reply GitCommands/AppTitleGenerator.cs]: I personally don't like `public static` properties. I don't like `public static` method either, but in this case it feels like a lesser evil. And we already have another object with the same behaviour.

I also entertained an idea of implementing a debug-only version of `IAppTitleGenerator` and inject it, but it felt like more work for little gain. Besides we haven't fleshed out the DI story (MEF could be the way to go, but I haven't figured how it fits with the testing story).

**Comment by gerhardol** [comment]
> I'm proposing to show the following:

This is fine for me, I just want something to differ between the installations I use.

If there are approvals, I will squash and merge in some days to await further comments.

**Comment by gerhardol** [comment]
@msftbot merge in 3 days

**Comment by RussKie** [comment]
I just realised I have broken a test that only runs in Debug... 😒

**Pull request merged**